coverlib/export_codecov_json.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package coverlib
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+)
+
+// ExportCodecovJson exports profile data to the Codecov custom coverage JSON
+// format (https://docs.codecov.com/docs/codecov-custom-coverage-format).
+//
+// Sample output (with added comments):
+//
+//	{
+//	  "coverage": {
+//	    "filename": {
+//	      "1": 0,      # line 1 missed
+//	      "2": 1,      # line 2 hit once
+//	      "7": 5       # line 7 hit 5 times
+//	    }
+//	  }
+//	}
+func ExportCodecovJson(p *Profiles, writer io.Writer) error {
+	// The output schema is odd, in that each line is a separate attribute
+	// (instead of being part of an array). This makes it hard to use Go's json
+	// machinery; we just produce the output directly.
```

Comment:
IMO, we should still use the json stdlib package. The output can be the JSON encoding of the following type (untested, but you get the point):

```go
type FileCoverage map[string]int // line number to hit count

type Data struct {
    Coverage map[string]FileCoverage `json:"coverage"` // filename to coverage
}
```

Reply:
Oh, awesome, I'll give that a shot!

Reply:
Done. I found in the docs that it even converts integer keys to strings automatically (unfortunately the resulting fields are still sorted as strings not integers).